### PR TITLE
hcp: give a proper error when using conflicting build name

### PR DESCRIPTION
**Pull request by mogrogan**
It was possible to put the same source 2 times in the build and when using HCP, it would error eventually since they are considered the samebuild from HCP side

### Test
```hcl
packer {
  required_plugins {
    docker = {
      version = ">= 1.0.8"
      source = "github.com/hashicorp/docker"
    }
  }
}

source "docker" "ubuntu-jammy" {
  image  = "ubuntu:jammy"
  commit = true
}

source "docker" "ubuntu-focal" {
  image  = "ubuntu:focal"
  commit = true
}

build {
  name = "jammy-build-block"

  hcp_packer_registry {
    bucket_name = "test-root"
    description = "hcp test with root"
  }
  sources = [
    "source.docker.ubuntu-jammy",
    "source.docker.ubuntu-jammy",
  ]
}
```

This would give this error:
```bash
$ packer build .
Error: Build name conflicts

When using HCP registry, build names are used as identifier. In this case, there
is 2 or more build called jammy-build-block.docker.ubuntu-jammy

```

### CODE REVIEW

File: internal/hcp/registry/hcl.go [hunk 1]
```diff
@@ -140,12 +140,14 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
 		diags = append(diags, bucketDiags...)
 	}
 
-	if diags.HasErrors() {
-		return nil, diags
+	registeringDiags := registerAllBuilds(&config.Builds, bucket)
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:143]: I know this will change, but in the current state, since `config.Builds[0]` is where we get the registry information from, this seems a bit weird to me, but I don't think it's necessarily wrong to do it now, just that it feels a bit off when looking at the code.
I wonder if we can add a small comment before doing this so as to clarify that the logic will be updated in the not too distant future, so that we have a trace in the code?

File: internal/hcp/registry/hcl_test.go [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+							SourceRef: hcl2template.SourceRef{
+								Type: "docker",
+								Name: "alpine",
+							},
+						},
+					},
+				},
+			},
+		},
+		"single build block with name and multiple sources": {
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl_test.go:71]: Just for clarity, I would add `err: false` to those test cases, I understand it's not strictly necessary as it's the zero-value, but it would make it clearer at first glance imho

File: internal/hcp/registry/hcl.go [hunk 3]
```diff
@@ -158,6 +160,38 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
 	}, nil
 }
 
+func registerAllBuilds(builds *hcl2template.Builds, bucket *Bucket) hcl.Diagnostics {
```
[the remark this answers is not in the document]
Comment (lbajolet-hashicorp) [reply internal/hcp/registry/hcl.go:163]: Looking at the tests, is this so you can invoke this besides `NewHCLRegistry`?

**Review by lbajolet-hashicorp — COMMENTED** [review]
Overall this looks good to me, I think it maintains existing compatibility, but to be sure, have you run HPATS on those?
I left a couple comments, but nothing blocking, I'll let you address those and come back for another round of review later.

File: internal/hcp/registry/hcl.go [hunk 4]
```diff
@@ -158,6 +160,38 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
...
+	for _, build := range *builds {
+		for _, source := range build.Sources {
+			name := prependIfNotEmpty(build.Name, source.String())
+
+			if _, alreadyExists := buildNameSet[name]; alreadyExists {
+				diags = append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Build name conflicts",
+					Detail: "When using HCP registry, build names are used as identifier. " +
+						"In this case, there is 2 or more build called " + name,
```
[the remark this answers is not in the document]
Comment (mogrogan) [reply internal/hcp/registry/hcl.go:176]: Currently, the only way to get a conflict is by using the same source on the build block. We probably should add a more precise message error once we support multiple build block

File: internal/hcp/registry/hcl.go [hunk 5]
```diff
@@ -144,20 +145,72 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
...
+	return diags
 }
 
 func (h *HCLRegistry) Metadata() Metadata {
 	return h.metadata
 }
+
+type fullBuildNames map[string]map[string]struct{}
+
+// set will return true if properly inserted and false if the values is conflicting
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:197]:
```suggestion
// set will return true if properly inserted and false if the values are conflicting
```

File: internal/hcp/registry/hcl.go [hunk 6]
```diff
@@ -144,20 +145,72 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
...
+		}
+	}
+	return diags
 }
 
 func (h *HCLRegistry) Metadata() Metadata {
 	return h.metadata
 }
+
+type fullBuildNames map[string]map[string]struct{}
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:195]: I'm not necessarily against the idea of introducing a structure for this, but it feels a bit verbous to get them, I wonder if having functions directly on `HCLRegistry` wouldn't be better for this purpose?

File: internal/hcp/registry/hcl.go [hunk 7]
```diff
@@ -18,10 +18,11 @@ import (
...
 // HCLRegistry is a HCP handler made for handling HCL configurations
 type HCLRegistry struct {
-	configuration *hcl2template.PackerConfig
-	bucket        *Bucket
-	ui            sdkpacker.Ui
-	metadata      *MetadataStore
+	configuration  *hcl2template.PackerConfig
+	bucket         *Bucket
+	ui             sdkpacker.Ui
+	metadata       *MetadataStore
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:24]: Come to think of it, since we change how we get the build names, does that impact that `MetadataStore`? It might not but I just want to be sure we considered delving into this before this gets merged, so that we don't break things later down the line

File: internal/hcp/registry/hcl.go [hunk 8]
```diff
@@ -144,20 +145,72 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
...
 }
 
 func (h *HCLRegistry) Metadata() Metadata {
 	return h.metadata
 }
+
+type fullBuildNames map[string]map[string]struct{}
+
+// set will return true if properly inserted and false if the values is conflicting
+func (b fullBuildNames) set(buildName string, sourceName string) bool {
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:198]: Assuming we remove that structure, this function is a one-time use thing, so the code might as well be rolled into the function that adds the builds I'd think

File: internal/hcp/registry/hcl.go [hunk 9]
```diff
@@ -144,20 +145,72 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
...
+	}
+	if _, ok := b[sourceName][buildName]; !ok {
+		b[sourceName][buildName] = struct{}{}
+		return true
+	}
+	return false
+}
+
+// get will return the properly formatted string taking name conflict into account
+func (b fullBuildNames) get(buildName string, sourceName string) string {
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:211]: Assuming this is rolled into the function (or not even), the signature could be changed to accept a `*CoreBuild` instead, so we can do the logic of getting the build from the source/build name combination if relevant

File: internal/hcp/registry/hcl.go [hunk 10]
```diff
@@ -144,20 +145,72 @@ func NewHCLRegistry(config *hcl2template.PackerConfig, ui sdkpacker.Ui) (*HCLReg
...
+	if len(b[sourceName]) > 1 && buildName != "" {
```
Comment (lbajolet-hashicorp) [inline internal/hcp/registry/hcl.go:212]: That might fail won't it? With the semantics we discussed, if we have a build block with a name, and only one source (so no conflict), we register the build using the name only, but this will return the build name and the source name?

Comment (mogrogan) [reply internal/hcp/registry/hcl.go:212]: Actually that check is just not required anymore since I inverted the way the nested set store names

**Review by lbajolet-hashicorp — APPROVED** [review]
Left two very minor nitpicks, but overall LGTM!
Pre-approving this now, we can re-approve once these are addressed, and merge this in

**Review by lbajolet-hashicorp — APPROVED** [review]
LGTM
